framework/DataObjects/DataSet.py

```diff
@@ -631,21 +633,25 @@ def write(self,fileName,style='netCDF',**kwargs):
     if style.lower() == 'netcdf':
       self._toNetCDF(fileName,**kwargs)
     elif style.lower() == 'csv':
-      if len(self._data[self.sampleTag])==0: #TODO what if it's just metadata?
+      if len(self.asDataset().variables)==0: #TODO what if it's just metadata?
         self.raiseAWarning('Nothing to write!')
         return
       #first write the CSV
       firstIndex = kwargs.get('firstIndex',0)
       self._toCSV(fileName,start=firstIndex,**kwargs)
       # then the metaxml
-      self._toCSVXML(fileName,**kwargs)
+      if 'DataSet' in self._meta.keys():
+        self._toCSVXML(fileName,**kwargs)
     # TODO dask?
     else:
       self.raiseAnError(NotImplementedError,'Unrecognized write style: "{}"'.format(style))
     if not self.hierarchical and 'RAVEN_isEnding' in self.getVars():
       return len(self._data.where(self._data['RAVEN_isEnding']==True,drop=True)['RAVEN_isEnding'])
     else:
-      return len(self) # so that other entities can track which realization we've written
+      if self.sampleTag in self._data.sizes.keys():
+        return len(self) # so that other entities can track which realization we've written
+      else:
+        return 0 # not 'RAVEN_sample_ID' is found
```

Comment:
Did we ever discuss whether all dataobjects should have a sample ID?  I think it's a discussion worth having, and I really don't know the correct answer.

Reply:
The ```RAVEN_sample_ID``` is added.